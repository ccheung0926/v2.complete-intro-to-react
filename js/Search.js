import React from 'react'
import ShowCard from './ShowCard'
import Header from './Header'
const { arrayOf, shape, string } = React.PropTypes

const Search = React.createClass({
  propTypes: {
    shows: arrayOf(shape({
      title: string,
      description: string
    }))
  },
  getInitialState () {
    return {
      searchTerm: ''
    }
  },
  handleSearchTermChange (event) {
    // use this.setState all time
    this.setState({searchTerm: event.target.value}) // <-- this.setState is an asynchronous functions that is going to schdule an update and eventually, it is going to empty that buffer into one setState, so only updating the bare minimum of things possible to keep your app performing
    // or this.state.searchTerm = event.target.value -- but this will cause the state and UI are not in sync, and the broswer does not re-render coz no one is watching
    // this.forceUpdate() -- update right now
  },
  render () {
    return (
      <div className='search'>
        {/* showSearch=true by default */}
        <Header
          showSearch
          searchTerm={this.state.searchTerm}
          handleSearchTermChange={this.handleSearchTermChange}
        />
        <div>
          {this.props.shows
            .filter((show) => `${show.title} ${show.description}`.toUpperCase().indexOf(this.state.searchTerm.toUpperCase()) >= 0)
            .map((show) => {
              return (
                <ShowCard key={show.imdbID} {...show} />
              )
            })}
        </div>
      </div>
    )
  }
})

export default Search
