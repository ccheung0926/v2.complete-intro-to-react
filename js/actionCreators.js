import { SET_SEARCH_TERM } from './actions'

export function setSearchTerm (searchTerm) {
{ /*{ searchTerm } = { searchTerm: searchTerm in ES6 }*/ }
  return { type: SET_SEARCH_TERM, searchTerm }
  }
}