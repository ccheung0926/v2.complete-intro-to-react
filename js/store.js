import { createStore } from 'redux'

import rootReducer from './reducers'

cost store = createStore(rootReducer)

export default store
