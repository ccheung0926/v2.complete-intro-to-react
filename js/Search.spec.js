import React from 'react'
import { Provider } from 'react-redux'
import { shallow, render } from 'enzyme'
import { shallowToJson } from 'enzyme-to-json'
import store from './store'
import { setSearchTerm } from './actionCreators'
import { Unwrapped as UnwrappedSearch } from './Search'
import ShowCard from './ShowCard'
import preload from '../public/data.json'

test('Search snapshot test', () => {
  const component = shallow(<UnwrappedSearch shows={preload.shows} searchTerm='' />)
  const tree = shallowToJson(component)
  expect(tree).toMatchSnapshot()
})

test('Search should render a ShowCard for each show', () => {
  const component = shallow(<Search />)
  expect(component.find(ShowCard).length).toEqual(preload.shows.length)
})

// test('Search should render correct amount of shows based on search', () => {
//   const searchWord = 'house'
//   const component = shallow(<Search />)
//   component.find('input').simulate('change', {target: {value: searchWord}})
// })

