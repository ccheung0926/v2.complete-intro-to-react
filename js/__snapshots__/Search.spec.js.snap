// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Search snapshot test 1`] = `
<div
  className="search"
>
  <header>
    <h1>
      svideo
    </h1>
    <input
      onChange={[Function]}
      placeholder="Search"
      type="text"
      value="this is the default string"
    />
  </header>
</div>
`;
